### PR TITLE
ostree: move admindir to /etc/alternatives.admindir

### PR DESCRIPTION
**Pull request by vrothberg**
`ostree container commit` wipes /var and thereby erases the data in /var/lib/alternatives; the directory used to store configs/symlinks of alternatives.

/var is not a good place for storing such configs since it won't receive updates on bootc images either.  We need to move to another location.

Hence, use /etc/alternatives.admindir when running on an ostree-based system unless /var/lib/alternatives is already present; a user may have worked around the problem.  This way we enable alternatives to work on ostree-based systems without breaking backwards compat.

Fixes: #9

### CODE REVIEW

**Comment by vrothberg** [comment]
Dockerfile to test:
```Dockerfile
FROM quay.io/fedora/fedora-bootc:40
COPY alternatives /usr/sbin/alternatives
RUN dnf -y install golang
RUN ls /etc/.alternatives.admindir
RUN go help > /dev/null
```

File: alternatives.8 [hunk 1]
```diff
@@ -214,6 +214,7 @@ A directory, by default
 containing
 .BR alternatives '
 state information.
+/etc/.alternatives.state on OSTree-based systems.
```
Comment (vrothberg) [inline alternatives.8:217]: Needs update.

File: chkconfig.spec [hunk 2]
```diff
@@ -81,9 +81,7 @@ mkdir -p $RPM_BUILD_ROOT/etc/chkconfig.d
 %{_sysconfdir}/chkconfig.d
 %{_sysconfdir}/init.d
 %{_sysconfdir}/rc.d
-%{_sysconfdir}/rc.d/init.d
 %{_sysconfdir}/rc[0-6].d
-%{_sysconfdir}/rc.d/rc[0-6].d
```
Comment (vrothberg) [inline chkconfig.spec:86]: These two need to be broken out into a dedicated commit.

File: alternatives.c [hunk 3]
```diff
@@ -1290,6 +1290,32 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
     return 0;
 }
 
+int dirExists(const char *path) {
+    struct stat stats;
+    stat(path, &stats);
```
Comment (cgwalters) [inline alternatives.c:1295]: This should probably handle errors (other than ENOENT). (That said of course this codebase doesn't do much proper error handling, I know)

File: alternatives.c [hunk 4]
```diff
@@ -1290,6 +1290,32 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
...
+
+    if (S_ISDIR(stats.st_mode))
```
Comment (cgwalters) [inline alternatives.c:1297]: I believe that at this point we'll be reading potentially uninitialized memory; I don't believe the `stat` system call (or C library wrapper) will necessarily initialize the target structure on error.

So we definitely want to bail early if we get an error.

Comment (vrothberg) [reply alternatives.c:1297]: Thanks for catching!

**Comment by travier** [comment]
I still think it would be better to do a [migration](https://github.com/fedora-sysv/chkconfig/pull/134#issuecomment-2250222781) as that would fix things for all systems once instead of us having to write migration scripts and ship them in Fedora CoreOS, Fedora Atomic Desktops, Fedora IoT, etc.

But this LGTM, although I've not tested it.

File: alternatives.c [hunk 5]
```diff
@@ -1290,6 +1290,41 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
     return 0;
 }
 
+int admindirExists(const char *path) {
```
Comment (lnykryn) [inline alternatives.c:1293]: I would name this as plane dirExists

File: alternatives.c [hunk 6]
```diff
@@ -1290,6 +1290,41 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
...
+    struct stat sb;
+    stat(path, &sb);
```
Comment (lnykryn) [inline alternatives.c:1295]: Not needed since we call it in the if statement below

File: alternatives.c [hunk 7]
```diff
@@ -1290,6 +1290,41 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
...
+
+    if (stat(path, &sb) || !S_ISDIR(sb.st_mode) || access(path, F_OK)) {
+        if (errno == ENOENT) {
+            return 0;
+        }
+        fprintf(stderr, _("admindir %s invalid\n"), path);
+        exit(2);
+    }
+    
+    if (S_ISDIR(sb.st_mode)) {
```
Comment (lnykryn) [inline alternatives.c:1305]: Not needed since this is already checked above

File: alternatives.c [hunk 8]
```diff
@@ -1290,6 +1290,41 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
...
+        return 1;
+    }
+
+    return 0;
+}
+
+int canUseAlternativeAdminDir() {
```
Comment (lnykryn) [inline alternatives.c:1312]: The name should reflect why we do the check some like isOstree, or canUseVar

File: alternatives.c [hunk 9]
```diff
@@ -1290,6 +1290,41 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
...
+    if (admindirExists("/var/lib/alternatives")) {
```
Comment (lnykryn) [inline alternatives.c:1313]: If there is only one line in the block we tend to not write the {} parenthesis around it.

File: alternatives.c [hunk 10]
```diff
@@ -1290,6 +1290,41 @@ static int listServices(const char *altDir, const char *stateDir, int flags) {
...
+        return 0;
+    }
+
+    if (fileExists("/run/ostree-booted")) {
```
Comment (lnykryn) [inline alternatives.c:1317]: Please merge this in one if.  fileExists("/run/ostree-booted") || isLink("/ostree")

File: alternatives.c [hunk 11]
```diff
@@ -1428,6 +1469,13 @@ int main(int argc, const char **argv) {
         exit(2);
     }
 
+    if ((statedir_changed == 0) && (stat(stateDir, &sb) == -1)) {
```
Comment (lnykryn) [inline alternatives.c:1472]: Maybe use the new dirExists function

**Review by lnykryn — CHANGES_REQUESTED** [review]

**Comment by vrothberg** [comment]
Dockerfile for testing ostree container:
```
FROM quay.io/fedora/fedora-bootc:40
COPY alternatives /usr/sbin/alternatives
RUN dnf -y install golang
RUN ls /etc/alternatives /etc/alternatives.admindir
RUN go help > /dev/null
```

Dockerfile to make sure that the admindir gets created even on package mode:
```
FROM fedora:40
COPY alternatives /usr/sbin/alternatives
RUN rm -rf /var/lib/alternatives
RUN dnf -y install golang
RUN ls /etc/alternatives /var/lib/alternatives
RUN go help > /dev/null
``` 

To test the systemd service, build the Dockerfule, run a detached container and podman-exec into it:
```
FROM quay.io/fedora/fedora-bootc:40
COPY ./systemd/alternatives-migration.service /usr/lib/systemd/system/
COPY alternatives-migration /usr/libexec/
```

File: Makefile [hunk 12]
```diff
@@ -9,7 +9,7 @@ PROG = chkconfig
 BINDIR = /sbin
 SBINDIR = /usr/sbin
 MANDIR = /usr/man
-ALTDIR = /var/lib/alternatives
+LIBEXECDIR = /usr/libexec
```
Comment (vrothberg) [inline Makefile:12]: Will revert the ALTDIR in the next push.

**Comment by lnykryn** [comment]
Honestly, I don't like the idea of moving those files on existing systems. It feels really fragile. If we want to move the db elsewhere, it should be done only on new installations and support reading from both directories.

**Comment by vrothberg** [comment]
> Honestly, I don't like the idea of moving those files on existing systems. It feels really fragile. If we want to move the db elsewhere, it should be done only on new installations and support reading from both directories.

I am happy to remove the migration part if you wish.  We decided against moving the DB on new installations as it may break existing users.  Let's say one machine upgrade to RHEL X.Y and another would do a fresh install, the paths would differ. We also wanted to make sure that user depending on /var/lib/alternatives (for whatever reason) are not impacted.

**Comment by vrothberg** [comment]
@travier @cgwalters can you chime in?

**Comment by cgwalters** [comment]
I think as far as a migration it seems doable across a RHEL major (and a fedora major)...the chance for breakage if we did it seems pretty low.

A migration would clearly be useful to be configurable - i.e. we can land the code, have it off by default and opt-in to start). So here's my proposal

- Add a mechanism to opt-in to the new location (`env ALTERNATIVES_STATEDIR=/etc/alternatives.admindir` or just `ALTERNATIVES_STATEDIR_NEW=1`? Or maybe again `mkdir /etc/alternatives.admindir` before installing)
- Change the code so that the new admindir, even if configured on is only used if the old `/var` path is nonexistent/inactive
- The migration code would be a separate opt-in (build-time option seems viable, then we can flip that on in just rhel10 and fedora 42's spec files *and/or* start by doing just the migration in e.g. fedora-bootc/FCOS/etc. to start)

**Comment by vrothberg** [comment]
> I think as far as a migration it seems doable across a RHEL major (and a fedora major)...the chance for breakage if we did it seems pretty low.
> 
> A migration would clearly be useful to be configurable - i.e. we can land the code, have it off by default and opt-in to start). So here's my proposal
> 
>     * Add a mechanism to opt-in to the new location (`env ALTERNATIVES_STATEDIR=/etc/alternatives.admindir` or just `ALTERNATIVES_STATEDIR_NEW=1`? Or maybe again `mkdir /etc/alternatives.admindir` before installing)

The latter would be my preference.

>     * Change the code so that the new admindir, even if configured on is only used if the old `/var` path is nonexistent/inactive

That would be doable.  The code still prefers /var if it exists..

>     * The migration code would be a separate opt-in (build-time option seems viable, then we can flip that on in just rhel10 and fedora 42's spec files _and/or_ start by doing just the migration in e.g. fedora-bootc/FCOS/etc. to start)

I think it's already opt-in by default by just shipping the system service but not enabling it. So fedora-bootc etc. can enable it if desired.

But I personally was already happy with previous PR who just did its magic on OSTree systems.  If we want to add more to it, let's dicuss and agree on the behavior.  I don't want to burn too much time.

**Comment by cgwalters** [comment]
Yeah, I hear you. I'm OK keeping this migration code around in a draft PR, and then we can continue to do the super targeted change?

But...I think the messy thing though is these are kind of entangled...if we change this code to detect `/run/ostree-booted`, in theory it could break non-bootc ostree systems where the user legitimately configured alternatives locally?

**Comment by vrothberg** [comment]
> Yeah, I hear you. I'm OK keeping this migration code around in a draft PR, and then we can continue to do the super targeted change?

Moved the current state into https://github.com/fedora-sysv/chkconfig/pull/139.

> But...I think the messy thing though is these are kind of entangled...if we change this code to detect `/run/ostree-booted`, in theory it could break non-bootc ostree systems where the user legitimately configured alternatives locally?

Can you elaborate?  The code will only use /etc/alternatives.admindir if /var/lib/alternatives doesn't exist, we are running on an OSTree host.  It won't switch if the --admindir flag is used.

File: alternatives.c [hunk 13]
```diff
@@ -1304,13 +1328,18 @@ int main(int argc, const char **argv) {
     struct stat sb;
     struct linkSet newSet = {NULL, NULL, NULL};
 
+    if (canUseAlternativeAdminDirUnderOSTree()) {
+        stateDir = "/etc/alternatives.admindir";
+    }
+
     setlocale(LC_ALL, "");
     bindtextdomain("chkconfig", "/usr/share/locale");
     textdomain("chkconfig");
```
Comment (ldv-alt) [inline alternatives.c:1337]: Since `canUseAlternativeAdminDirUnderOSTree()` is invoked before `setlocale()`, the diagnostic message will not be translated. Also, if `--admindir` option is used, `canUseAlternativeAdminDirUnderOSTree()` should not be invoked at all.  You can even create a setup where even `--help` and `--version` options will not work after this change.

File: alternatives.c [hunk 14]
```diff
@@ -1304,13 +1328,18 @@ int main(int argc, const char **argv) {
...
 
     if (!argv[1])
         return usage(2);
 
+    int statedir_changed = 0;
```
Comment (ldv-alt) [inline alternatives.c:1342]: There is no real need for a new variable.  I'd suggest initializing `stateDir` to NULL and check it later instead.

Comment (vrothberg) [reply alternatives.c:1342]: Thanks!

File: alternatives.c (hunk 13 above)
Comment (vrothberg) [reply alternatives.c:1337]: Will update, thanks for catching!

**Comment by vrothberg** [comment]
Good to go from my end.
@cgwalters @travier @lnykryn @ldv-alt WDYT?

Heads up: I'll be on PTO next week.

**Comment by lnykryn** [comment]
I am OK with the concept. However, I still want to debug what is going on with https://github.com/fedora-sysv/chkconfig/pull/135/commits/6d6b96f770bcae40e3159e4062abb40446c7e1a6, and some small bits in the code bother me. But to save everyone's time I can just force-push that here.

**Comment by vrothberg** [comment]
> I am OK with the concept. However, I still want to debug what is going on with [6d6b96f](https://github.com/fedora-sysv/chkconfig/commit/6d6b96f770bcae40e3159e4062abb40446c7e1a6), and some small bits in the code bother me. But to save everyone's time I can just force-push that here.

Sounds great, thanks!  Feel free to take over this PR.

File: chkconfig.spec [hunk 15]
```diff
@@ -95,11 +95,12 @@ mkdir -p $RPM_BUILD_ROOT/etc/chkconfig.d
 %files -n alternatives
 %license COPYING
 %dir /etc/alternatives
+%ghost /etc/alternatives.admindir
```
Comment (dtardon) [inline chkconfig.spec:98]: This still needs to be `%dir`. That is:
```
%ghost %dir /etc/alternatives.admindir
```
and the same for the next line.

**Review by dtardon — CHANGES_REQUESTED** [review]

File: chkconfig.spec (hunk 15 above)
Comment (lnykryn) [reply chkconfig.spec:98]: good catch, but I found out it is not enough, it also needs attr for -qV to be happy

File: alternatives.c [hunk 16]
```diff
@@ -1454,12 +1466,26 @@ int main(int argc, const char **argv) {
...
+        stateDir = "/var/lib/alternatives";
+        if(!dirExists(stateDir)) {
+            if(isOSTree())
+                stateDir="/etc/alternatives-admindir";
+
+            if (mkdir(stateDir, 0755) < 0 && errno != EEXIST) {
+                fprintf(stderr, _("failed creating admindir: %s\n"), strerror(errno));
+                exit(2);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (lnykryn) [reply alternatives.c:1487]: I already talked about the acccess and I like performant code :-)

File: alternatives.c [hunk 17]
```diff
@@ -1454,12 +1466,26 @@ int main(int argc, const char **argv) {
...
+    // and /etc/alternatives-admindir on OSTree systems, if the dir does not exist, we will create it
+    // if the stateDir is explicitly set, we will *not* try to create the dir and fail immediately if it does not exist
+    if (!stateDir) {
+        stateDir = "/var/lib/alternatives";
+        if (!dirExists(stateDir)) {
+            if (isOSTree())
+                stateDir = "/etc/alternatives-admindir";
+
+            if (mkdir(stateDir, 0755) < 0 && errno != EEXIST) {
+                fprintf(stderr, _("failed creating admindir: %s\n"), strerror(errno));
```
Comment (ldv-alt) [inline alternatives.c:1484]: grammar: please either `error creating` or `failed to create`

File: alternatives.c [hunk 18]
```diff
@@ -1454,12 +1466,26 @@ int main(int argc, const char **argv) {
         }
     }
 
-    if (stat(altDir, &sb) || !S_ISDIR(sb.st_mode) || access(altDir, F_OK)) {
+    if (!dirExists(altDir)) {
```
[the remark this answers is not in the document]
Comment (dtardon) [reply alternatives.c:1469]: True... But then it'd be better to remove it in a separate commit.

**Review by dtardon — APPROVED** [review]
LGTM

**Comment by vrothberg** [comment]
Thanks!  Do we need to cut a new upstream release to get the changes into Fedora/RHEL?

**Comment by lnykryn** [comment]
@jamacku already did that https://github.com/fedora-sysv/chkconfig/commit/81cda82ee74c97efd0e879b340e3daeeed10654f
